core/slice.go

```diff
@@ -586,69 +587,70 @@ func (sl *Slice) computePendingHeader(localPendingHeaderWithTermini types.Pendin
 }
 
-// writePhCache dom writes a given pendingHeaderWithTermini to the cache with the terminus used as the key.
-func (sl *Slice) writeToPhCache(pendingHeaderWithTermini types.PendingHeader) {
-	sl.phCache[pendingHeaderWithTermini.Termini[terminiIndex]] = pendingHeaderWithTermini
-}
-
-// pickPhCacheHead determines if the provided pendingHeader should be selected and returns true if selected
-func (sl *Slice) pickPhCacheHead(reorg bool, externPendingHeaderWithTermini types.PendingHeader, domOrigin bool) bool {
-	if reorg {
-		sl.pendingHeaderHeadHash = externPendingHeaderWithTermini.Termini[terminiIndex]
-		return true
-	}
-
-	localPendingHeader, exists := sl.phCache[sl.pendingHeaderHeadHash]
-
-	if domOrigin {
-		//calc local cache head reorg
-		localCacheReorg := true
-		for i := 0; i < common.NodeLocation.Context(); i++ {
-			localCacheReorg = (externPendingHeaderWithTermini.Header.NumberArray()[i].Cmp(localPendingHeader.Header.NumberArray()[i]) >= 0) && localCacheReorg
-		}
-
-		if exists && localCacheReorg && (externPendingHeaderWithTermini.Header.NumberU64() > localPendingHeader.Header.NumberU64()) {
-			return sl.updateCurrentPendingHeader(externPendingHeaderWithTermini)
-		}
-	}
-
-	return false
-}
-
 // updateCurrentPendingHeader compares the externPh parent td to the sl.pendingHeader parent td and sets sl.pendingHeader to the exterPh if the td is greater
-func (sl *Slice) updateCurrentPendingHeader(externPendingHeader types.PendingHeader) bool {
+func (sl *Slice) updatePendingHeader(externPendingHeader types.PendingHeader, local bool) bool {
 	externTd := sl.hc.GetTdByHash(externPendingHeader.Header.ParentHash())
 	currentTd := sl.hc.GetTdByHash(sl.phCache[sl.pendingHeaderHeadHash].Header.ParentHash())
 	log.Debug("updateCurrentPendingHeader:", "currentParent:", sl.phCache[sl.pendingHeaderHeadHash].Header.ParentHash(), "currentTd:", currentTd, "externParent:", externPendingHeader.Header.ParentHash(), "externTd:", externTd)
 	if currentTd != nil && externTd != nil {
-		if currentTd.Cmp(externTd) < 0 {
-			sl.pendingHeaderHeadHash = externPendingHeader.Termini[terminiIndex]
+		if local {
+			if currentTd.Cmp(externTd) < 0 {
+				return true
+			}
+		} else {
+			if currentTd.Cmp(externTd) <= 0 {
+				return true
+			}
 		}
 	} else {
 		log.Warn("updateCurrentPendingHeader:", "currentParent:", sl.phCache[sl.pendingHeaderHeadHash].Header.ParentHash(), "currentTd:", currentTd, "externParent:", externPendingHeader.Header.ParentHash(), "externTd:", externTd)
 		return false
 	}
-	return true
+	return false
+}
+
+// writePhCache dom writes a given pendingHeaderWithTermini to the cache with the terminus used as the key.
+func (sl *Slice) writeToPhCache(pendingHeaderWithTermini types.PendingHeader, local bool, terminiIndex int) {
+	_, exist := sl.phCache[pendingHeaderWithTermini.Termini[terminiIndex]]
+	if !exist {
+		sl.phCache[pendingHeaderWithTermini.Termini[terminiIndex]] = pendingHeaderWithTermini
```

Comment:
This is a shallow copy. Does this need to be a deep copy?

Reply:
This should be fine as in both cases the use of this function is instance variables.  However, I will add a deep copy to keep its usage safe.